src/tester.rs

```diff
@@ -37,52 +38,128 @@ pub fn run_test(run: &mut Command, testcase: &ClashTestCase) -> Result<TestRunRe
     } else {
-        Ok(TestRunResult::RuntimeError { stderr })
+        Ok(TestRunResult::RuntimeError { stdout, stderr })
     }
 }
 
 pub fn show_test_result(result: &TestRunResult, testcase: &ClashTestCase) {
-    // TODO: use OutputStyle
-    let success_style = Style::new().on(Color::Green);
-    let failure_style = Style::new().on(Color::Red);
-    let error_style = Style::new().on(Color::Red);
-    let title_style = Style::new().fg(Color::Yellow);
-    let stderr_style = Style::new().fg(Color::Red);
-    let out_style = Style::new().fg(Color::White);
-    let ws_style = Some(Style::new().fg(Color::RGB(43, 43, 43)));
-
-    let title = title_style.paint(&testcase.title);
+    // Create class Tester?
+    let style = TestCaseStyle::default();
+    let formatter = Formatter::default();
+
+    let title = style.title.paint(&testcase.title);
     match result {
         TestRunResult::Success => {
-            println!("{} {}", success_style.paint("PASS"), title);
+            println!("{} {}", style.success.paint("PASS"), title);
         }
 
         TestRunResult::WrongOutput { stderr, stdout } => {
-            println!("{} {}", failure_style.paint("FAIL"), title);
+            println!("{} {}", style.failure.paint("FAIL"), title);
             if !stderr.is_empty() {
-                println!("{}", stderr_style.paint(stderr.trim_end()));
+                println!("{}", style.stderr.paint(stderr.trim_end()));
             }
-            let formatter = Formatter::default();
-            println!(
-                "==== EXPECTED ====\n{}",
-                formatter.show_whitespace(&testcase.test_out, &out_style, &ws_style)
-            );
-            println!(
-                "===== ACTUAL =====\n{}\n",
-                formatter.show_whitespace(stdout, &out_style, &ws_style)
-            );
+
+            let input = &testcase.test_in;
+            let expected = &testcase.test_out;
+            let received = &stdout;
+            // compare in bulk
+            print_pair(input, expected, received, &formatter, &style);
+            println!("");
+            // compare but stop printing RECEIVED when an error is found (+format)
+            compare(input, expected, received, &formatter, &style);
         }
 
-        TestRunResult::RuntimeError { stderr } => {
-            println!("{} {}", error_style.paint("ERROR"), title);
-            println!("{}\n", stderr_style.paint(stderr.trim_end()));
+        TestRunResult::RuntimeError { stdout, stderr } => {
+            println!("{} {}", style.error.paint("ERROR"), title);
+            println!("{}\n", style.stderr.paint(stderr.trim_end()));
+            print_pair(&testcase.test_in,&testcase.test_out, &stdout, &formatter, &style);
         }
     }
 }
+
+fn compare(input: &str, expected: &str, received: &str, formatter: &Formatter, style: &TestCaseStyle) {
+    let errors_allowed = 10;
+
+    // If nothing was received at all, special message?
```

Comment:
I'm not sure about the special message, it would need to somehow be abundantly clear that Nothing doesn't mean that they literally printed "Nothing".

Reply:
At the moment "Nothing" is colored (in the foreground) in red so no confusion is possible. Another option that I've seen in `atcoder-tools test` it's just a newline.

Reply:
When looking at the text the user has no reference for what it "normally" looks like so it's still easy to get confused even if the color is different (especially if it's red like an error would be). The faint whitespace color might work better? It could be a good idea to also change the text from "Nothing" to "(no output)" to reduce the risk of it getting confused with any programming language specific types or errors (such as `Nothing` from Haskell).